File: leetcode/python/18.4Sum.py
class Main:

  @staticmethod
  def fourSum(nums,target:int):
    result = []
    if len(nums)<4:
      return result
    nums.sort()
    maximum = nums[-1]
    for i in range(len(nums)):
      if i > 0 and nums[i]==nums[i-1]:
        continue
      if nums[i] + 3 * maximum < target:
        continue
      if nums[i]*4 > target:
        break
      for j in range(i+1,len(nums)):
        if j > i+1 and nums[j]==nums[j-1]:
          continue
        start = j+1
        end= len(nums)-1
        while start < end :
          total = nums[i]+nums[j]+nums[start]+nums[end]
          if total == target :
            tmp = []
            tmp.append(nums[i])
            tmp.append(nums[j])
            tmp.append(nums[start])
            tmp.append(nums[end])
            result.append(tmp)
            startVal = nums[start]
            endVal = nums[end]
            while start < end and startVal == nums[start]:
              start += 1
            while start < end and endVal == nums[end]:
              end -= 1
          elif total < target:
            start+=1
          else :
            end -=1
    return result

print(Main.fourSum([1,0,-1,0,-2,2],0))
              

            

File: leetcode/c++/18.4Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> fourSum(vector<int> nums, int target)
{
    vector<vector<int>> result;
    if (nums.size() == 0)
    {
        return result;
    }
    sort(nums.begin(), nums.end());
    int max = nums[nums.size() - 1];
    for (int i = 0; i < nums.size(); i++)
    {
        if (i > 0 && nums[i] == nums[i - 1])
        {
            continue;
        }
        if (nums[i] + 3 * max < target)
        {
            continue;
        }
        if (nums[i] * 4 > target)
        {
            break;
        }
        for (int j = i + 1; j < nums.size(); j++)
        {
            if (j > i + 1 && nums[j] == nums[j - 1])
            {
                continue;
            }
            int start = j + 1;
            int end = nums.size() - 1;
            while (start < end)
            {
                int sum = nums[i] + nums[j] + nums[start] + nums[end];
                if (sum == target)
                {
                    vector<int> tmp;
                    tmp.push_back(nums[i]);
                    tmp.push_back(nums[j]);
                    tmp.push_back(nums[start]);
                    tmp.push_back(nums[end]);
                    result.push_back(tmp);
                    int startVal = nums[start];
                    int endVal = nums[end];
                    while (start < end && startVal == nums[start])
                    {
                        start++;
                    }
                    while (start < end && endVal == nums[end])
                    {
                        end--;
                    }
                }
                else if (sum < target)
                {
                    start++;
                }
                else
                {
                    end--;
                }
            }
        }
    }
    return result;
}

int main()
{
    vector<int> v{1, 0, -1, 0, -2, 2};
    for (auto i : fourSum(v, 0))
    {
        for (auto j : i)
        {
            cout << j << " ";
        }
        cout << endl;
    };
    return 0;
}